File: step-1-basic-model/src/domain/order/OrderRepository.ts
import { Order } from './Order';
import { OrderId } from './OrderId';
import { CustomerId } from '../customer/CustomerId';

/**
 * OrderRepository Interface
 * Aggregate Root単位でRepositoryを定義
 */
export interface OrderRepository {
  save(order: Order): Promise<void>;
  findById(id: OrderId): Promise<Order | null>;
  findByCustomerId(customerId: CustomerId): Promise<Order[]>;
  delete(id: OrderId): Promise<void>;
}

File: python-ddd-sample/domain/order/order_repository.py
"""
OrderRepository Interface
Aggregate Root単位でRepositoryを定義
"""
from abc import ABC, abstractmethod
from typing import List, Optional
from .order import Order
from .order_id import OrderId
from ..customer.customer_id import CustomerId


class OrderRepository(ABC):
    """注文リポジトリのインターフェース"""
    
    @abstractmethod
    async def save(self, order: Order) -> None:
        """注文を保存"""
        pass
    
    @abstractmethod
    async def find_by_id(self, order_id: OrderId) -> Optional[Order]:
        """IDで注文を検索"""
        pass
    
    @abstractmethod
    async def find_by_customer_id(self, customer_id: CustomerId) -> List[Order]:
        """顧客IDで注文を検索"""
        pass
    
    @abstractmethod
    async def delete(self, order_id: OrderId) -> None:
        """注文を削除"""
        pass